File: apps/api/prompt_piper_api/routes/sessions.py
from uuid import UUID

from fastapi import APIRouter, Depends, Response, status

from prompt_piper_api.config import get_settings
from prompt_piper_api.llm.factory import create_llm_client_from_env
from prompt_piper_api.schemas.inference import SendToInferenceRequest, SendToInferenceResponse
from prompt_piper_api.schemas.precision import (
    ApplyPrecisionReplacementRequest,
    PrecisionReviewResponse,
    PrecisionSuggestRequest,
    PrecisionSuggestResponse,
)
from prompt_piper_api.schemas.session import (
    AnswerClarificationRequest,
    AskTheLocalsResponse,
    ClarificationSuggestionsResponse,
    CreateSessionFromTemplateRequest,
    CreateSessionRequest,
    EditDraftRequest,
    FinalizeSessionRequest,
    GenerateArtifactsRequest,
    SessionDetailResponse,
    to_session_detail,
    to_session_response,
)
from prompt_piper_api.services.artifact_factory import create_artifact_export_service
from prompt_piper_api.services.audit_log_service import AuditLogService
from prompt_piper_api.services.external_inference_service import ExternalInferenceService
from prompt_piper_api.services.git_registry_service import GitRegistryService
from prompt_piper_api.services.optimization.engine import TokenOptimizationEngine
from prompt_piper_api.services.session_service import SessionService
from prompt_piper_api.services.session_store import create_session_store
from prompt_piper_api.services.similarity_factory import create_similarity_check_service

router = APIRouter(prefix="/sessions", tags=["sessions"])

_session_service: SessionService | None = None


def get_session_service() -> SessionService:
    global _session_service
    if _session_service is None:
        settings = get_settings()
        audit = AuditLogService(settings.audit_log_path)
        artifact_export = create_artifact_export_service(settings)
        _session_service = SessionService(
            llm=create_llm_client_from_env(),
            registry=GitRegistryService(settings.registry_path),
            similarity=create_similarity_check_service(settings),
            optimizer=TokenOptimizationEngine(),
            artifact_export=artifact_export,
            external_inference=ExternalInferenceService(
                settings,
                audit,
                settings.artifacts_path,
            ),
            audit=audit,
            store=create_session_store(settings.sessions_path),
        )
    return _session_service


@router.post("", response_model=SessionDetailResponse, status_code=201)
def create_session(
    payload: CreateSessionRequest,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.create_session(
        initial_request=payload.initial_request,
        title=payload.title,
    )
    return to_session_response(result)


@router.post("/{session_id}/template", response_model=SessionDetailResponse, status_code=201)
def create_session_from_template(
    session_id: UUID,
    payload: CreateSessionFromTemplateRequest | None = None,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.create_session_from_template(
        session_id,
        title=None if payload is None else payload.title,
    )
    return to_session_response(result)


@router.get("/{session_id}", response_model=SessionDetailResponse)
def get_session(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    record = service.get_session(session_id)
    return to_session_detail(record)


@router.post("/{session_id}/delete", status_code=status.HTTP_204_NO_CONTENT)
def delete_session_post(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> Response:
    service.delete_session(session_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.delete("/{session_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_session(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> Response:
    service.delete_session(session_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{session_id}/clarify/suggest", response_model=ClarificationSuggestionsResponse)
def suggest_clarification(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> ClarificationSuggestionsResponse:
    result = service.suggest_clarification(session_id)
    return ClarificationSuggestionsResponse.model_validate(result.model_dump())


@router.post("/{session_id}/clarify/locals", response_model=AskTheLocalsResponse)
def ask_the_locals(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> AskTheLocalsResponse:
    result = service.ask_the_locals(session_id)
    return AskTheLocalsResponse.model_validate(result.model_dump())


@router.post("/{session_id}/answer", response_model=SessionDetailResponse)
def answer_clarification(
    session_id: UUID,
    payload: AnswerClarificationRequest,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.answer_clarification(session_id, payload.answer)
    return to_session_response(result)


@router.post("/{session_id}/clarify/complete", response_model=SessionDetailResponse)
def complete_clarification(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.complete_clarification(session_id)
    return to_session_response(result)


@router.post("/{session_id}/edit", response_model=SessionDetailResponse)
def edit_session_draft(
    session_id: UUID,
    payload: EditDraftRequest,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.edit_draft(
        session_id, payload.instruction, body=payload.body
    )
    return to_session_response(result)


@router.post("/{session_id}/workflow/reopen/edit", response_model=SessionDetailResponse)
def reopen_session_for_edit(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.reopen_for_edit(session_id)
    return to_session_response(result)


@router.post("/{session_id}/workflow/rerun/similarity", response_model=SessionDetailResponse)
def rerun_session_similarity_check(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.rerun_similarity_check(session_id)
    return to_session_response(result)


@router.post("/{session_id}/workflow/rerun/optimize", response_model=SessionDetailResponse)
def rerun_session_optimization(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.rerun_optimization(session_id)
    return to_session_response(result)


@router.post("/{session_id}/finalize", response_model=SessionDetailResponse)
def finalize_session(
    session_id: UUID,
    payload: FinalizeSessionRequest | None = None,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    body = payload or FinalizeSessionRequest()
    result = service.finalize(
        session_id,
        acknowledge_first_shot_risk=body.acknowledge_first_shot_risk,
    )
    return to_session_response(result)


@router.post("/{session_id}/optimize", response_model=SessionDetailResponse)
def optimize_session(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.optimize(session_id)
    return to_session_response(result)


@router.post("/{session_id}/optimize/approve", response_model=SessionDetailResponse)
def approve_optimization(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.approve_optimization(session_id)
    return to_session_response(result)


@router.get("/{session_id}/precision", response_model=PrecisionReviewResponse)
def get_precision_review(
    session_id: UUID,
    service: SessionService = Depends(get_session_service),
) -> PrecisionReviewResponse:
    return service.get_precision_review(session_id)


@router.post("/{session_id}/precision/suggest", response_model=PrecisionSuggestResponse)
def suggest_precision_replacement(
    session_id: UUID,
    payload: PrecisionSuggestRequest,
    service: SessionService = Depends(get_session_service),
) -> PrecisionSuggestResponse:
    result = service.suggest_precision_replacement(session_id, finding_id=payload.finding_id)
    return PrecisionSuggestResponse.from_service(result)


@router.post("/{session_id}/precision/apply", response_model=SessionDetailResponse)
def apply_precision_replacement(
    session_id: UUID,
    payload: ApplyPrecisionReplacementRequest,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.apply_precision_replacement(
        session_id,
        finding_id=payload.finding_id,
        replacement=payload.replacement,
    )
    return to_session_response(result)


@router.post("/{session_id}/artifacts", response_model=SessionDetailResponse)
def generate_session_artifacts(
    session_id: UUID,
    payload: GenerateArtifactsRequest | None = None,
    service: SessionService = Depends(get_session_service),
) -> SessionDetailResponse:
    result = service.generate_artifacts(
        session_id,
        include_pdf=True if payload is None else payload.include_pdf,
        export_folder_label=None if payload is None else payload.export_folder_label,
    )
    return to_session_response(result)


@router.post("/{session_id}/send-to-inference", response_model=SendToInferenceResponse)
def send_session_to_inference(
    session_id: UUID,
    payload: SendToInferenceRequest,
    service: SessionService = Depends(get_session_service),
) -> SendToInferenceResponse:
    result = service.send_to_inference(
        session_id,
        explicit_approval=payload.explicit_approval,
        api_endpoint_id=payload.api_endpoint_id,
    )
    return SendToInferenceResponse(
        provider=result.provider,
        model=result.model,
        prompt_id=result.prompt_id,
        version=result.version,
        timestamp=result.timestamp,
        artifact_location=result.artifact_location,
        inference_response_artifact_path=result.inference_response_artifact_path,
        response_text=result.response_text,
    )

File: apps/web/src/api/sessions.ts
import type { HealthResponse, LlmHealthResponse } from "@prompt-piper/shared";
import { apiFetch } from "./http";
import type {
  AskTheLocalsResponse,
  ClarificationSuggestionsResponse,
  InferenceSettingsResponse,
  PrecisionReviewResponse,
  PrecisionSuggestResponse,
  RegistryPromptDetail,
  RegistryPromptSummary,
  SendToInferenceResponse,
  SessionDetailResponse,
} from "./types";

export function fetchHealth(): Promise<HealthResponse> {
  return apiFetch<HealthResponse>("/health");
}

export function fetchLlmHealth(): Promise<LlmHealthResponse> {
  return apiFetch<LlmHealthResponse>("/health/llm");
}

export function createSession(payload: {
  initial_request: string;
  title?: string;
}): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>("/sessions", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function createSessionFromTemplate(
  sessionId: string,
  title?: string,
): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/template`, {
    method: "POST",
    body: JSON.stringify(title ? { title } : {}),
  });
}

export function fetchSession(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}`);
}

export function deleteSession(sessionId: string): Promise<void> {
  return apiFetch<void>(`/sessions/${sessionId}/delete`, { method: "POST" });
}

export function answerClarification(
  sessionId: string,
  answer: string,
): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/answer`, {
    method: "POST",
    body: JSON.stringify({ answer }),
  });
}

export function completeClarification(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/clarify/complete`, {
    method: "POST",
  });
}

export function suggestClarification(
  sessionId: string,
): Promise<ClarificationSuggestionsResponse> {
  return apiFetch<ClarificationSuggestionsResponse>(`/sessions/${sessionId}/clarify/suggest`, {
    method: "POST",
  });
}

export function askTheLocals(sessionId: string): Promise<AskTheLocalsResponse> {
  return apiFetch<AskTheLocalsResponse>(`/sessions/${sessionId}/clarify/locals`, {
    method: "POST",
  });
}

export function editDraft(
  sessionId: string,
  payload: { instruction: string } | { body: string },
): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/edit`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function finalizeSession(
  sessionId: string,
  options?: { acknowledgeFirstShotRisk?: boolean },
): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/finalize`, {
    method: "POST",
    body: JSON.stringify({
      acknowledge_first_shot_risk: options?.acknowledgeFirstShotRisk ?? false,
    }),
  });
}

export function reopenForEdit(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/workflow/reopen/edit`, {
    method: "POST",
  });
}

export function rerunSimilarityCheck(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/workflow/rerun/similarity`, {
    method: "POST",
  });
}

export function rerunOptimization(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/workflow/rerun/optimize`, {
    method: "POST",
  });
}

export function optimizeSession(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/optimize`, {
    method: "POST",
  });
}

export function approveOptimization(sessionId: string): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/optimize/approve`, {
    method: "POST",
  });
}

export function fetchPrecisionReview(sessionId: string): Promise<PrecisionReviewResponse> {
  return apiFetch<PrecisionReviewResponse>(`/sessions/${sessionId}/precision`);
}

export function suggestPrecisionReplacement(
  sessionId: string,
  findingId: string,
): Promise<PrecisionSuggestResponse> {
  return apiFetch<PrecisionSuggestResponse>(`/sessions/${sessionId}/precision/suggest`, {
    method: "POST",
    body: JSON.stringify({ finding_id: findingId }),
  });
}

export function applyPrecisionReplacement(
  sessionId: string,
  findingId: string,
  replacement: string,
): Promise<SessionDetailResponse> {
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/precision/apply`, {
    method: "POST",
    body: JSON.stringify({ finding_id: findingId, replacement }),
  });
}

export function generateArtifacts(
  sessionId: string,
  options?: { includePdf?: boolean; exportFolderLabel?: string },
): Promise<SessionDetailResponse> {
  const includePdf = options?.includePdf ?? true;
  const payload: { include_pdf: boolean; export_folder_label?: string } = {
    include_pdf: includePdf,
  };
  const label = options?.exportFolderLabel?.trim();
  if (label) {
    payload.export_folder_label = label;
  }
  return apiFetch<SessionDetailResponse>(`/sessions/${sessionId}/artifacts`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function fetchRegistryPrompts(): Promise<RegistryPromptSummary[]> {
  return apiFetch<RegistryPromptSummary[]>("/registry/prompts");
}

export function fetchRegistryPrompt(promptId: string): Promise<RegistryPromptDetail> {
  return apiFetch<RegistryPromptDetail>(`/registry/prompts/${encodeURIComponent(promptId)}`);
}

export function fetchInferenceSettings(): Promise<InferenceSettingsResponse> {
  return apiFetch<InferenceSettingsResponse>("/settings/inference");
}

export function sendToInference(
  sessionId: string,
  options?: { apiEndpointId?: string | null },
): Promise<SendToInferenceResponse> {
  return apiFetch<SendToInferenceResponse>(`/sessions/${sessionId}/send-to-inference`, {
    method: "POST",
    body: JSON.stringify({
      explicit_approval: true,
      api_endpoint_id: options?.apiEndpointId ?? null,
    }),
  });
}
